src/web/tokenserver.rs

```diff
@@ -1,26 +1,165 @@
+    replaced_at: Option<i64>,
+    #[sql_type = "Text"]
+    node: String,
+    #[sql_type = "Nullable<Bigint>"]
+    keys_changed_at: Option<i64>,
+}
+
+#[derive(serde::Serialize)]
+pub struct TokenServerResult {
+    id: String,
+    key: String,
+    uid: String,
+    api_endpoint: String,
+    duration: String,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct Claims {
+    pub sub: String,
+    pub iat: i64,
+    pub exp: i64,
+}
 
 pub fn get(
-    request: TokenServerRequest,
+    auth: BearerAuth,
 ) -> impl Future<Output = Result<HttpResponse, BlockingError<ApiError>>> {
-    block(move || get_sync(request).map_err(Into::into)).map_ok(move |_result| {
-        // TODO turn _result into a json response.
+    block(move || get_sync(&auth).map_err(Into::into)).map_ok(move |result| {
         HttpResponse::Ok()
             .content_type("application/json")
-            .body("{}")
+            .body(serde_json::to_string(&result).unwrap())
     })
 }
 
-pub fn get_sync(_request: TokenServerRequest) -> Result<TokenServerResult, ApiError> {
-    // TODO Perform any blocking calls needed to respond to the tokenserver request.
-    Ok(TokenServerResult {})
+pub fn get_sync(auth: &BearerAuth) -> Result<TokenServerResult, ApiError> {
+    // the public rsa components come from
+    // https://oauth.accounts.firefox.com/v1/jwks
+    // TODO we should fetch it from an environment var instead of hardcoding it.
+    let token_data = decode::<Claims>(
+        &auth.token(),
+        &DecodingKey::from_rsa_components("2lDphW0lNZ4w1m9CfmIhC1AxYG9iwihxBdQZo7_6e0TBAi8_TNaoHHI90G9n5d8BQQnNcF4j2vOs006zlXcqGrP27b49KkN3FmbcOMovvfesMseghaqXqqFLALL9us3Wstt_fV_qV7ceRcJq5Hd_Mq85qUgYSfb9qp0vyePb26KEGy4cwO7c9nCna1a_i5rzUEJu6bAtcLS5obSvmsOOpTLHXojKKOnC4LRC3osdR6AU6v3UObKgJlkk_-8LmPhQZqOXiI_TdBpNiw6G_-eishg8V_poPlAnLNd8mfZBam-_7CdUS4-YoOvJZfYjIoboOuVmUrBjogFyDo72EPTReQ", "AQAB"),
+        &Validation::new(Algorithm::RS256),
+    ).map_err(|ee| {
+        ApiError::from(ApiErrorKind::Internal(format!("Unable to decode token_data: {:}", ee)))
+    })?;
+    println!("token data! {:?}", token_data);
+    let email = format!("{:}@api.accounts.firefox.com", token_data.claims.sub);
+
+    // TODO pull out of settings instead
+    let shared_secret = env::var("SYNC_MASTER_SECRET").expect("SYNC_MASTER_SECRET must be set");
+    let database_url =
+        env::var("TOKENSERVER_DATABASE_URL").expect("TOKENSERVER_DATABASE_URL must be set");
+
+    let connection = MysqlConnection::establish(&database_url)
+        .unwrap_or_else(|_| panic!("Error connecting to {}", database_url));
+    let user_record = diesel::sql_query(
+        "select users.uid, services.pattern, users.email, users.generation, users.client_state, users.created_at, users.replaced_at, nodes.node, users.keys_changed_at from users, services, nodes where users.email = ? and services.id = users.service and nodes.id = users.nodeid and nodes.service = services.id")
```

Comment:
nit: Might want to format this a bit. I know pjenvey is an advocate of https://www.sqlstyle.guide/, and I'll admit it makes parsing what's going on here a bit easier.